### PR TITLE
Freshness checks factories

### PR DESCRIPTION
**Pull request by dpeng817**
Adds two factory methods to produce freshness checks from a list of passed-in assets. One factory for time window partitioned assets, and one for non partitioned assets.
More than likely, in non-time-window cases, users will want a check per partition, which is currently not possible but will soon be possible with the multi_asset_check decorator.
Originally, this PR had both of these within one factory, which proved to be very hard to parse.

I opted to go for only a cron_schedule parameter for the partitioned version, and make it required. The cron schedule is the most flexible parameterization, since it allows for completely independent control of the freshness condition w.r.t. the underlying partitions, but also allows you to account for things like daylight savings. The reason it's required is that without it, the underlying partition will always be late immediately after the time window ends, which is confusing behavior.

The one thing I'm not too happy about is that we can only catch certain errors with the passed in errors (ie, partitioned asset passed into non partitioned method) at runtime. But I don't think this needs to block implementation.

How I tested this
Added unit tests for the following cases
- All possible parameterizations / making sure definition-time error paths are exercised
- Combinations of cron and lag params for non partitioned assets
- Cron + partitioned assets
- Both observation and materialization events
- Incorrect assets passed to builders, and the runtime errors that ensue.

### CODE REVIEW

**Comment by dpeng817** [comment]
* **#20377** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/20377?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by sryza — COMMENTED** [review]
> An interesting question is; what do we do when a user passes in max_lag_minutes alongside partitioned assets. Do we error, do we ignore it? For now I just choose to ignore it.

This makes me suspect we should have two separate methods.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    The freshness_cron parameter can be used to influence the time window that is checked. For example, if I have a daily partitioned asset,
+    and I specify a freshness_cron for 12:00 PM, and I run the check at 1:00 AM March 12, I will check for the time window
+    March 10 12:00 AM - March 11 12:00 AM, since the most recent completed tick of the cron was at 12:00 PM March 11.
+    For partitioned assets, the maximum_lag_minutes parameter is not used.
+
+    Args:
+        assets (Sequence[Union[CoercibleToAssetKey, AssetsDefinition, SourceAsset]): The assets to construct checks for. For each passed in
+            asset, there will be a corresponding constructed `AssetChecksDefinition`.
+        freshness_cron (Optional[str]): The cron for which freshness is defined.
+        maximum_lag_minutes (Optional[int]): The maximum lag in minutes that is acceptable for the asset.
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks.py:54]: Similar to the suggestion above: is there some way of phrasing this in terms of "The check will pass if"?

**Review by sryza — CHANGES_REQUESTED** [review]
Haven't made it entirely through yet, but logging some initial comments

File: python_modules/dagster/dagster/_core/definitions/freshness_checks.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    assets: Sequence[Union[CoercibleToAssetKey, AssetsDefinition, SourceAsset]],
+    freshness_cron: Optional[str] = None,
+    maximum_lag_minutes: Optional[int] = None,
+    severity: Optional[AssetCheckSeverity] = None,
+) -> Sequence[AssetChecksDefinition]:
+    """For each provided asset, constructs a freshness check definition.
+
+    An asset is considered fresh if it has been materialized or observed within a certain time window.
+    The freshness_cron and maximum_lag_minutes parameters are used to define this time window.
+    freshness_cron defines the time at which the asset should have been materialized, and maximum_lag_minutes provides a tolerance for the range at which the asset can arrive.
```
[the remark this answers is not in the document]
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks.py:36]: I just figured out how to add a ruler to vscode. Hopefully that helps catch these earlier going forward.

**Comment by sryza** [comment]
I just pushed some changes:
- Support observations in addition to materializations
- Add some metadata
- Correct / add description
- Rebase on top of Sean's AssetChecksDefinition refactor and AssetLayer refactor

File: python_modules/dagster/dagster/_core/definitions/freshness_checks.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+from .asset_checks import AssetChecksDefinition
+from .assets import AssetsDefinition, SourceAsset
+from .events import AssetKey, CoercibleToAssetKey
+from .partition import PartitionsDefinition
+from .time_window_partitions import TimeWindowPartitionsDefinition
+
+DEFAULT_FRESHNESS_SEVERITY = AssetCheckSeverity.WARN
+
+
+def build_freshness_checks(
```
[the remark this answers is not in the document]
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks.py:27]: Maybe it's just me but I like to make something public in a separate PR so that it's easily walked back, but not a huge deal either way.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+from .utils import (
+    DEFAULT_FRESHNESS_SEVERITY,
+    asset_to_keys_iterable,
+    ensure_no_duplicate_assets,
+    get_last_updated_timestamp,
+    retrieve_latest_record,
+)
+
+
+def build_non_partitioned_freshness_checks(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py:28]: Naming thing (which I think applies to the partitioned check factory below too): in the nearish future, checks themselves will be able to be partitioned (i.e. pass / fail per partition instead of per the entire asset).  This name implies that the check will not be partitioned, not that the asset is not partitioned.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_window_partitioned.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    For example, let's say I have a daily time-window partitioned asset, and I specify an offset of
+    60 minutes. This means that I expect March 11 to March 12 data to be seen by March 13 1:00 AM.
+    If the data is not seen, the asset check will not pass.
+
+    Args:
+        assets (Sequence[Union[CoercibleToAssetKey, AssetsDefinition, SourceAsset]): The assets to
+            construct checks for. For each passed in asset, there will be a corresponding
+            constructed `AssetChecksDefinition`.
+        offset_minutes (int): The number of minutes to allow for the asset to be considered fresh. Defaults to 0.
+
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/time_window_partitioned.py:47]: Unnecessary extra newline

File: python_modules/dagster/dagster_tests/definitions_tests/freshness_checks_tests/test_time_window_partitioned.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        assets=[my_asset],
+        offset_minutes=10,
+    )
+
+    freeze_datetime = start_time
+    with pendulum.test(freeze_datetime):
+        # With no events, check fails.
+        assert_check_result(my_asset, instance, freshness_checks, AssetCheckSeverity.WARN, False)
+
+        # Add an event for an old partition. Still fails
```
Comment (sryza) [inline python_modules/dagster/dagster_tests/definitions_tests/freshness_checks_tests/test_time_window_partitioned.py:141]: To test that the offset works as expected, I would expect to see an assert that this passes with `pendulum.test(pendulum.datetime(2021, 1, 1, 0, 9, 0, tz="UTC"))`.  I.e. the offset tolerates us being 10 minutes late, so if we're only 9 minutes past the start of the day, then we're fine.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    *,
+    assets: Sequence[Union[CoercibleToAssetKey, AssetsDefinition, SourceAsset]],
+    freshness_cron: Optional[str] = None,
+    maximum_lag_minutes: Optional[int] = None,
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py:32]: Should `maximum_lag_minutes` be required? Having trouble conceptualizing how this works without it.

**Review by sryza — COMMENTED** [review]
My first reaction to `offset_minutes` was that it makes a lot of sense and seems easier to deal with than a cron.  E.g. for an hourly partitioned asset where you want to make sure that the data arrives 10 minutes after the start of the hour, `offset_minutes` seems like the most straightforward way to specify this.

When I thought about it a little more though, I realized that the cron is more flexible, as well as I think more intuitive in the daily case. E.g. imagine that, for a daily-partitioned asset, you want to make sure that data arrives by 9 am new york time every day.  You wouldn't be able to express this with an offset, because the offset is different depending on daylight savings time.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_window_partitioned.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+) -> Sequence[AssetChecksDefinition]:
+    """For each provided time-window partitioned asset, constructs a freshness check definition.
+
+    Will emit a runtime error for any assets that are not time-window partitioned.
+
+    An asset is considered fresh if its most recent time window is materialized or observed within
+    the specified offset.
+
+    For example, let's say I have a daily time-window partitioned asset, and I specify an offset of
+    60 minutes. This means that I expect March 11 to March 12 data to be seen by March 13 1:00 AM.
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/time_window_partitioned.py:39]: to be seen by March 12 1:00 AM, right?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    ]
+
+
+def _build_freshness_check_for_asset(
+    asset: Union[CoercibleToAssetKey, AssetsDefinition, SourceAsset],
+    freshness_cron: Optional[str],
+    maximum_lag_minutes: Optional[int],
+    severity: AssetCheckSeverity,
+) -> Sequence[AssetChecksDefinition]:
+    from ..decorators.asset_check_decorator import asset_check
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py:94]: Why import this down here?

**Comment by dpeng817** [comment]
@sryza couldn't you express "9am New York time" as a number of minutes past the current partition? I was thinking maybe the timezone piece you couldn't, but a simple cron string doesn't provide a timezone anyway. I do see what you mean about it being a bit more expressive, so down to add it if you feel strongly. Wondering if we want to have both or just one of them

**Comment by sryza** [comment]
The issue is that on the day of the daylight savings spring forward transition, 9 hours after midnight is 10 am. And on the day of the fall back transition, 9 hours after midnight is 8 am.

**Comment by dpeng817** [comment]
ah, good point.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py (hunk 7 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py:32]: I buy being maximally restrictive with the parameterizations. I did add a test, but I agree it's kinda weird.

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_window_partitioned.py (hunk 8 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/time_window_partitioned.py]: that's right

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        for check in _build_freshness_check_for_asset(
+            asset,
+            freshness_cron=freshness_cron,
+            maximum_lag_minutes=maximum_lag_minutes,
+            severity=severity,
+        )
+    ]
+
+
+def _build_freshness_check_for_asset(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py:93]: I think `_build_freshness_checks_for_assets` would be more accurate?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+                }
+
+                if passed:
+                    description = None
+                else:
+                    if latest_cron_tick is None:
+                        description = f"Last update was more than {maximum_lag_minutes} minutes ago"
+                    else:
+                        description = (
+                            f"Last update was more than {maximum_lag_minutes} before last cron tick"
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py:141]:
```suggestion
                            f"Last update was more than {maximum_lag_minutes} minutes before last cron tick"
```

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+            return AssetCheckResult(
+                passed=passed, severity=severity, description=description, metadata=metadata
+            )
+
+        checks.append(the_check)
+
+    return checks
+
+
+def get_latest_completed_cron_tick(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py:159]: This seems generally useful enough that it should go in the same utils file as `reverse_cron_string_iterator`

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+def build_freshness_checks_for_non_partitioned_assets(
+    *,
+    assets: Sequence[Union[CoercibleToAssetKey, AssetsDefinition, SourceAsset]],
+    maximum_lag_minutes: int,
+    freshness_cron: Optional[str] = None,
+    severity: AssetCheckSeverity = DEFAULT_FRESHNESS_SEVERITY,
+) -> Sequence[AssetChecksDefinition]:
+    """For each provided asset, constructs a freshness check definition.
+
+    An asset is considered fresh if it has been materialized or observed within a certain time
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py:38]: ❤️ ❤️ ❤️  the line wrapping

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+    maximum_lag_minutes provides a tolerance for the range at which the asset can arrive.
+
+    Let's say an asset kicks off materializing at 12:00 PM and takes 10 minutes to complete.
+    Allowing for operational constraints and delays, the asset should always be materialized by
+    12:30 PM. 12:00 PM provides the lower boundary, since the asset kicks off no earlier than this
+    time. Then, we set freshness_cron to "30 12 * * *", which means the asset is expected by
+    12:30 PM, and maximum_lag_minutes to 30, which means the asset can be materialized no earlier
+    than 12:00 PM.
+
+    A time-window partitioned asset is considered fresh if the latest completed time window is
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/non_partitioned.py:50]: This should go on the other method, right?

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_window_partitioned.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+                    description = (
+                        f"Partition {expected_partition_key} arrived on time. "
+                        f"Materialized on {last_updated_time.to_formatted_date_string()} at "
+                        f"{last_updated_time.to_time_string()}."
+                    )
+                else:
+                    description = (
+                        f"Partition {expected_partition_key} arrived on time. "
+                        f"Latest data was from "
+                        f"{last_updated_time.to_formatted_date_string()} at "
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/freshness_checks/time_window_partitioned.py:140]: No strong preference, but could also use use `to_formatted_datetime_string` I believe.

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/time_window_partitioned.py:140]: Something to be careful with here is timezone. When testing out similar code, I ended up in a confusing situation where this string was in a different timezone than all the other timestamps in the UI. IMO we should either:
- Show the timezone in this string
- Rely on the metadata values to show the datetime, because the frontend can use the same timezone for it as for all the other timestamps

**Review by sryza — APPROVED** [review]
Left a few comments worth heeding, but otherwise looks great!

File: python_modules/dagster/dagster/_core/definitions/freshness_checks/time_window_partitioned.py (hunk 15 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/_core/definitions/freshness_checks/time_window_partitioned.py:140]: For consistency's sake I agree it's less goofy to keep it out of the description.

**Comment by schrockn** [comment]
@sryza @dpeng817  are `build_freshness_checks_for_non_partitioned_assets` and `build_freshness_checks_for_time_window_partitioned_assets` going to both be public-facing APIs? We can instead have a generic thing that handles arbitrary partition schemes or at least composes better?

**Comment by dpeng817** [comment]
@schrockn open to ideas, but we found a single generic API to actually be more confusing.
The maximum_lag_parameter only makes sense for non-partitioned assets. The definition of freshness is also different depending on if the assets are time-window partitioned, non-partitioned, or partitioned using other schemes, and we felt this polluted the freshness_cron parameter,  as its meaning was different depending on the type of the asset.

All of this led to a massive and unwieldly docstring that we felt was somewhat unacceptable, which led us to this alternative. It is verbose, true, but the complexity is well encapsulated by the API, IMO.

The only other idea I have is allowing users to pass one generic list of assets, and then allowing users to parameterize different types of assets with different objects. Something like:
```python
build_freshness_checks(
    assets=[non_partitioned_asset, partitioned_asset],
    partitioned_params=PartitionedFreshnessParams(...),
    non_partitioned_params=NonPartitionedFreshnessParams(...),
)
```
At which point, we could also just allow people to provide a lambda which will, at runtime (since that's the only time we can determine the partitioning of an asset for sure), convert an assetsdefinition+asset key to a set of freshness params
```python
def build_params(assets_def: AssetsDefinition, asset_key: AssetKey):
    if assets_def.partitions_def:
        return PartitionedFreshnessParams(...)
    return NonPartitionedFreshnessParams(...)

build_freshness_checks(
     assets=[non_partitioned_asset, partitioned_asset],
     param_builder=build_params,
)
```

Again, open to ideas tho, because the verbosity is a bit painful.

**Comment by sryza** [comment]
@dpeng817 - what would the docstring look like in that case?

**Comment by sryza** [comment]
A concern I have about that approach is: if I understand correctly, it seems like both the check parameters _and_ the check success criteria are defined by the XParams class. Ie the XParams class is really the sole owner of the check definition logic. So it's not really clear what the role of build_freshness_checks is other than to do isinstance checks and dispatch to the subclasses?

**Comment by dpeng817** [comment]
I think that's basically right. Also controls metadata so we can schedule running the checks too.
